### PR TITLE
Widen return values in CallTailCallTarget stubs

### LINKED ISSUE #55253: Test Failure:Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests.TestBaseline_TestWithTwoParams_Success
Run [runtime-coreclr libraries-jitstress 20210706.1](https://dev.azure.com/dnceng/public/_build/results?buildId=1221394&view=ms.vss-test-web.build-test-results-tab&runId=36396784&resultId=102882&paneView=debug)

Failed tests:
```
net6.0-windows-Release-arm64-CoreCLR_checked-tailcallstress-Windows.10.Arm64v8.Open

- Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests.TestBaseline_TestWithTwoParams_Success(argumentList: \"0, LogLevel.Error, \\\"M0 {a1} {a2}\\\", SkipEnabledCh\"...)
- Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests.TestBaseline_TestWithTwoParams_Success(argumentList: \"0, LogLevel.Error, \\\"M0 {a1} {a2}\\\"\")
- Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests.TestBaseline_TestWithTwoParams_Success(argumentList: \"eventId: 0, level: LogLevel.Error, message: \\\"M0 {\"...)
- Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests.TestBaseline_TestWithNestedClass_Success
- Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests.TestBaseline_TestWithSkipEnabledCheck_Success
```
**One of error messages**
```
Line 9 does not match.
Expected Line:
global::Microsoft.Extensions.Logging.LoggerMessage.Define(global::Microsoft.Extensions.Logging.LogLevel.Information, new global::Microsoft.Extensions.Logging.EventId(0, nameof(M0)), "Message: When using SkipEnabledCheck, the generated code skips logger.IsEnabled(logLevel) check before calling log. To be used when consumer has already guarded logger method in an IsEnabled check.", true);
Actual Line:
global::Microsoft.Extensions.Logging.LoggerMessage.Define(global::Microsoft.Extensions.Logging.LogLevel.Information, new global::Microsoft.Extensions.Logging.EventId(-256, nameof(M0)), "Message: When using SkipEnabledCheck, the generated code skips logger.IsEnabled(logLevel) check before calling log. To be used when consumer has already guarded logger method in an IsEnabled check.", true);
Expected: True
Actual:   False


Stack trace
   at Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests.VerifyAgainstBaselineUsingFile(String filename, String testSourceCode) in /_/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorEmitterTests.cs:line 164
   at Microsoft.Extensions.Logging.Generators.Tests.LoggerMessageGeneratorEmitterTests.TestBaseline_TestWithSkipEnabledCheck_Success() in /_/src/libraries/Microsoft.Extensions.Logging.Abstractions/tests/Microsoft.Extensions.Logging.Generators.Tests/LoggerMessageGeneratorEmitterTests.cs:line 49
--- End of stack trace from previous location ---
```

### PR DESCRIPTION
**Pull request by jakobbotsch**
In the managed calling convention the callee is responsible for widening
the return value up to 4 bytes. We need to mimic this in portable
tailcall helpers as the JIT cannot know which call eventually produced
the value that was stored in the return value buffer passed to the
dispatcher.

Fix #55253

### CODE REVIEW

**Comment by jakobbotsch** [comment]
CI is running at https://dev.azure.com/dnceng/public/_build/results?buildId=1239894&view=results (normal) and https://dev.azure.com/dnceng/public/_build/results?buildId=1239922&view=results (jitstress).

**Comment by jakobbotsch** [comment]
For a little context, we only just recently started making use of the widening done in callees in #54864 which exposed this bug in portable tailcall helpers.

The test failures seem to be unrelated test infra failures. @jkotas PTAL.

File: src/coreclr/vm/tailcallhelp.cpp [hunk 1]
```diff
@@ -579,7 +601,20 @@ MethodDesc* TailCallHelp::CreateCallTargetStub(const TailCallInfo& info)
...
         pCode->EmitLDARG(ARG_RET_VAL);
         pCode->EmitLDLOC(resultLcl);
-        EmitStoreTyHnd(pCode, info.RetTyHnd);
+        bool isSigned;
+        if (NeedsReturnWidening(info.RetTyHnd, &isSigned))
+        {
+            if (isSigned)
+                pCode->EmitCONV_I4();
+            else
+                pCode->EmitCONV_U4();
```
Comment (jakobbotsch) [inline src/coreclr/vm/tailcallhelp.cpp:610]: Is this even needed?

Comment (jkotas) [reply src/coreclr/vm/tailcallhelp.cpp:610]: It should not be. `EmitLDLOC` should do the appropriate widening implicitly.

**Comment by jkotas** [comment]
Why does the slow tail call path need to know about this? 

If the JIT wants to play calling convention tricks for fast tail calls, it is fine with me. But I think it would be preferable to keep the slow tail call path simple.

**Comment by jakobbotsch** [comment]
The alternative is to reject explicit tail calls when there is implicit widening. That's what the logic has been previously.

I took a look and ECMA-335 does seem to imply that tail. is permissible whenever the callee's return type is assignable-to the caller's return type, which covers these newly allowed scenarios if I'm not wrong.

I don't have a strong opinion. It should not be difficult to reject explicit tailcalls when types are mismatching on the JIT side. Do you think that's better?

**Comment by jkotas** [comment]
Yes, I think it would be better to reject explicit tailcalls when the return types do not match exactly, like it was done before.

The platform-specific (or CoreCLR/RyuJIT specific) calling convention details should not be leaking into when tailcalls work or do not work.

**Comment by jakobbotsch** [comment]
Superseded by #55989.

**Pull request closed without merging**